### PR TITLE
(1049) Add assess view referral routes

### PR DESCRIPTION
**Pull request by yndajas**
## Context

Submitted referrals are currently viewable by `ACP_REFERRER` users in Refer

## Changes in this PR

- Allow `ACP_PROGRAMME_TEAM` users to view referrals in Assess

I'm not too sure about the integration testing approach here. Happy to talk it through if there's a cleaner approach to doing this without over-duplicating tests (in terms of number of tests the suite needs to run more than the code - the code could still be DRY even if we test every scenario with both roles)

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the Accredited Programmes API for this change to work...
  - [ ] ... and they have been released to production already

### Post-merge

- [ ] This adds, extends or meaningfully modifies a feature...
  - [ ] ... and I have written or updated an end-to-end test for the happy path in the [Accredited Programmes E2E repo](https://github.com/ministryofjustice/hmpps-accredited-programmes-e2e)
- [ ] This makes new expectations of the API...
  - [ ] ... and I have notified the API developer(s) of changes to the contract tests (Pact), or the API is already compliant
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

**Comment by yndajas** [comment]
[Add `assess` routes for viewing submitted referrals (S)](https://trello.com/c/kTJBGJIT/1049-add-assess-routes-for-viewing-submitted-referrals-s)

File: integration_tests/e2e/showReferral.cy.ts [hunk 1]
```diff
@@ -66,13 +67,15 @@ context('Viewing a submitted referral', () => {
     }),
     addedByDisplayName: StringUtils.convertToTitleCase(user.name),
   }
+  let randomRole: 'ACP_PROGRAMME_TEAM' | 'ACP_REFERRER'
+  let paths: typeof assessPaths | typeof referPaths
 
   beforeEach(() => {
+    randomRole = Math.random() < 0.5 ? 'ACP_PROGRAMME_TEAM' : 'ACP_REFERRER'
```
Comment (jsrobertson) [inline integration_tests/e2e/showReferral.cy.ts:74]: Is this preferable to using `faker.helpers.arrayElement(['ACP_PROGRAMME_TEAM', 'ACP_REFERRER'])` like we do elsewhere?

**Review by jsrobertson — APPROVED** [review]
Nice one

**Review by Gweaton — COMMENTED** [review]
I'm not 100% sure on the shared tests assess & refer myself. I'd prefer for us to duplicate the tests to allow them to serve as more effective documentation e.g.

Only a PT user should be able to view Assess, only a Referrer user should be able to view Refer.

Having each journey documented as its own integration test would help. I think readability is probably preferred over DRYness, but if there's a way we can share some of this implementation across both journeys, that would be good too.

File: integration_tests/e2e/showReferral.cy.ts (hunk 1 above)
Comment (yndajas) [reply integration_tests/e2e/showReferral.cy.ts:74]: Ha, that would've been smarter. I think I'm just not used to thinking in Faker on the integration test side!

**Comment by yndajas** [comment]
> I'm not 100% sure on the shared tests assess & refer myself. I'd prefer for us to duplicate the tests to allow them to serve as more effective documentation e.g.
> 
> Only a PT user should be able to view Assess, only a Referrer user should be able to view Refer.
> 
> Having each journey documented as its own integration test would help. I think readability is probably preferred over DRYness, but if there's a way we can share some of this implementation across both journeys, that would be good too.

Yeah, I was more thinking in terms of integration test suite speed than DRYness of code, but maybe we should test each scenario

File: integration_tests/support/sharedTests.ts [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        sentenceInformationPage.shouldContainNavigation(path)
+        sentenceInformationPage.shouldContainBackLink('#')
+        sentenceInformationPage.shouldContainCourseOfferingSummaryList(coursePresenter, organisation.name)
+        sentenceInformationPage.shouldContainSubmissionSummaryList(referral)
+        sentenceInformationPage.shouldContainSubmittedReferralSideNavigation(path, referral.id)
+        sentenceInformationPage.shouldContainImportedFromText()
+        sentenceInformationPage.shouldContainSentenceDetailsSummaryCard()
+      },
+
+      setup: (role: ApplicationRole): void => {
```
Comment (yndajas) [inline integration_tests/support/sharedTests.ts:199]: Could call this `beforeEach`

File: integration_tests/e2eReferDisabled/assess.cy.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
+import { ApplicationRoles } from '../../server/middleware/roleBasedAccessMiddleware'
```
Comment (yndajas) [inline integration_tests/e2eReferDisabled/assess.cy.ts:1]: We could add these test to shared too but I'm not as bothered

File: integration_tests/e2e/refer/show.cy.ts [hunk 4]
```diff
@@ -1,211 +1,36 @@
...
-      })
-      personalDetailsPage.shouldHavePersonDetails(person)
-      personalDetailsPage.shouldContainNavigation(path)
-      personalDetailsPage.shouldContainBackLink('#')
-      personalDetailsPage.shouldContainCourseOfferingSummaryList(coursePresenter, organisation.name)
-      personalDetailsPage.shouldContainSubmissionSummaryList(referral)
-      personalDetailsPage.shouldContainSubmittedReferralSideNavigation(path, referral.id)
-      personalDetailsPage.shouldContainImportedFromText()
-      personalDetailsPage.shouldContainPersonalDetailsSummaryCard()
+      sharedTests.referrals.show.personalDetails(ApplicationRoles.ACP_REFERRER)
```
Comment (Gweaton) [inline integration_tests/e2e/refer/show.cy.ts:7]: Can we make this something like `sharedTests.referrals.showsPersonalDetailsPage` to keep this in an assertion-style format?

File: integration_tests/e2e/refer/show.cy.ts [hunk 5]
```diff
@@ -1,211 +1,36 @@
...
-        programmeHistoryPage.shouldContainNavigation(path)
-        programmeHistoryPage.shouldContainBackLink('#')
-        programmeHistoryPage.shouldContainCourseOfferingSummaryList(coursePresenter, organisation.name)
-        programmeHistoryPage.shouldContainSubmissionSummaryList(referral)
-        programmeHistoryPage.shouldContainSubmittedReferralSideNavigation(path, referral.id)
-        programmeHistoryPage.shouldContainHistorySummaryCards(courseParticipationsPresenter, referral.id, {
-          change: false,
-          remove: false,
-        })
+        sharedTests.referrals.show.programmeHistory.some(ApplicationRoles.ACP_REFERRER)
```
Comment (Gweaton) [inline integration_tests/e2e/refer/show.cy.ts:14]: `showsProgrammeHistoryPage`

File: integration_tests/e2e/refer/show.cy.ts [hunk 6]
```diff
@@ -1,211 +1,36 @@
...
-        })
-
-        programmeHistoryPage.shouldHavePersonDetails(person)
-        programmeHistoryPage.shouldContainNavigation(path)
-        programmeHistoryPage.shouldContainBackLink('#')
-        programmeHistoryPage.shouldContainCourseOfferingSummaryList(coursePresenter, organisation.name)
-        programmeHistoryPage.shouldContainSubmissionSummaryList(referral)
-        programmeHistoryPage.shouldContainSubmittedReferralSideNavigation(path, referral.id)
-        programmeHistoryPage.shouldContainNoHistoryParagraph()
+        sharedTests.referrals.show.programmeHistory.none(ApplicationRoles.ACP_REFERRER)
```
Comment (Gweaton) [inline integration_tests/e2e/refer/show.cy.ts:20]: `showsEmptyProgrammeHistoryPage`

File: integration_tests/e2e/refer/show.cy.ts [hunk 7]
```diff
@@ -1,211 +1,36 @@
...
-      })
-      sentenceInformationPage.shouldHavePersonDetails(person)
-      sentenceInformationPage.shouldContainNavigation(path)
-      sentenceInformationPage.shouldContainBackLink('#')
-      sentenceInformationPage.shouldContainCourseOfferingSummaryList(coursePresenter, organisation.name)
-      sentenceInformationPage.shouldContainSubmissionSummaryList(referral)
-      sentenceInformationPage.shouldContainSubmittedReferralSideNavigation(path, referral.id)
-      sentenceInformationPage.shouldContainImportedFromText()
-      sentenceInformationPage.shouldContainSentenceDetailsSummaryCard()
+      sharedTests.referrals.show.sentenceInformation(ApplicationRoles.ACP_REFERRER)
```
Comment (Gweaton) [inline integration_tests/e2e/refer/show.cy.ts:27]: `showsSentenceInformationPage`

**Review by Gweaton — COMMENTED** [review]
One last thing!

**Review by Gweaton — APPROVED** [review]
Looking good, tests are clear and easy to read, thanks for all the changes.

**Pull request merged**